File: java/me/aov/sellgui/listeners/PlayerLeaveListener.java
package me.aov.sellgui.listeners;

import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.ItemStack;
import me.aov.sellgui.SellGUIMain;
import me.aov.sellgui.gui.PriceEvaluationGUI;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class PlayerLeaveListener implements Listener {

    private final SellGUIMain plugin;
    private static final Map<UUID, ItemStack> pendingItems = new HashMap<>();

    public PlayerLeaveListener(SellGUIMain plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        UUID playerId = event.getPlayer().getUniqueId();

        ItemStack pendingItem = pendingItems.remove(playerId);
        if (pendingItem != null) {

            if (event.getPlayer().getInventory().firstEmpty() != -1) {
                event.getPlayer().getInventory().addItem(pendingItem);
            } else {

                event.getPlayer().getWorld().dropItemNaturally(event.getPlayer().getLocation(), pendingItem);
            }

            plugin.getLogger().info("Returned evaluation item to " + event.getPlayer().getName() + " on quit");
        }
    }

    public static void storePendingItem(UUID playerId, ItemStack item) {
        if (item != null) {
            pendingItems.put(playerId, item.clone());
        }
    }

    public static void removePendingItem(UUID playerId) {
        pendingItems.remove(playerId);
    }

    public static boolean hasPendingItem(UUID playerId) {
        return pendingItems.containsKey(playerId);
    }
}

File: src/main/kotlin/me/aov/sellgui/listeners/PlayerLeaveListener.kt
package me.aov.sellgui.listeners

import me.aov.sellgui.SellGUIMain
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.inventory.ItemStack
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class PlayerLeaveListener(private val plugin: SellGUIMain) : Listener {
    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        val player = event.player
        val pendingItem = pendingItems.remove(player.uniqueId) ?: return
        if (player.inventory.firstEmpty() != -1) {
            player.inventory.addItem(pendingItem)
        } else {
            player.world.dropItemNaturally(player.location, pendingItem)
        }
        plugin.logger.info("Returned evaluation item to ${player.name} on quit")
    }

    companion object {
        private val pendingItems = ConcurrentHashMap<UUID, ItemStack>()

        @JvmStatic
        fun storePendingItem(playerId: UUID, item: ItemStack?) {
            if (item != null) pendingItems[playerId] = item.clone()
        }

        @JvmStatic
        fun removePendingItem(playerId: UUID) {
            pendingItems.remove(playerId)
        }

        @JvmStatic
        fun hasPendingItem(playerId: UUID): Boolean = pendingItems.containsKey(playerId)
    }
}
